master/buildbot/wamp/connector.py

```diff
@@ -0,0 +1,140 @@
+# This file is part of .  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright  Team Members
+
+from autobahn.twisted.wamp import ApplicationSession
+from autobahn.twisted.wamp import Service
+from autobahn.wamp.exception import TransportLost
+
+from twisted.internet import defer
+from twisted.python import failure
+from twisted.python import log
+
+from buildbot.util import service
+
+
+class MasterService(ApplicationSession, service.AsyncMultiService):
+
+    """
+    concatenation of all the wamp services of buildbot
+    """
+
+    def __init__(self, config):
+        ApplicationSession.__init__(self)
+        service.AsyncMultiService.__init__(self)
+        self.config = config
+        self.setServiceParent(config.extra['parent'])
+
+    @defer.inlineCallbacks
+    def onJoin(self, details):
+        log.msg("Wamp connection suceed!")
+        for handler in [self] + self.services:
+            yield self.register(handler)
+            yield self.subscribe(handler)
+        yield self.publish("org.buildbot.%s.connected" % (self.master.masterid))
+        self.parent.service = self
+        self.parent.serviceDeferred.callback(self)
+
+    @defer.inlineCallbacks
+    def onLeave(self, details):
+        # XXX We don't handle crossbar reboot, or any other disconnection well.
+        # this is a tricky problem, as we would have to reconnect with expononential backoff
+        # re-subscribe to subscriptions, queue messages until reconnection.
+        # This is quite complicated, and I believe much better handled in autobahn
+        # It is possible that such failure is practically non-existant
```

Comment:
FWIW, re-connection logic *is* coming in Autobahn, after an API for arbitrary transports for ApplicationRunner is figured out. If you're interested in this, see https://github.com/tavendo/AutobahnPython/issues/409 and also issue 295.